### ISSUE TITLE
Deprecate isLiteralPoolOnDemandOn in the z Systems codegen

### ISSUE THREAD

**Issue by fjeremic** [eclipse/omr#939]
The `isLiteralPoolOnDemandOn` is a z Systems particular codegen query. Historically it was used to lock a particular register and designate it to hold the literal pool address throughout the lifetime of the compilation.

On modern z/Architecture, more specifically with the introduction of load relative instructions in z10, we have the ability to load values from the literal pool with a single relative load. The relative loads, LRL for example, are RIL type instructions meaning we can address a literal pool that it at most 4GB away. With the deprecation of tiered code cache it is inconceivable that the literal pool cannot be addressed with relative load instructions. As such, we always benefit from freeing up the literal pool register for assignment as the register pressure decreases at no cost to our ability to load values from the literal pool.

In fact if the `DynamicLiteralPool` optimization runs, which is the case in pretty much most optimization strategies we will always free the literal pool register. As such I propose we deprecate this the ability to lock the literal pool register and prevent it from being assigned as it frankly does not provide any benefit currently.

I would like to hear some opinions or alternative viewpoints on this issue from subject matter experts. @joransiu @ebadyano @NigelYiboYu any concerns regarding this?

**Comment by lmaisons** [comment]
Pardon my ignorance, but how did tiered code cache affect this?  As far as I know, code caches were never anywhere near 4G large, and thus anywhere in both hot and cold areas should have always been reachable within 4G.

I guess I'm worried your thinking instead of trampolines, which could potentially still be needed.

**Comment by fjeremic** [comment]
@lmaisons if I'm understanding this correctly tiered code caches were code caches that were allocated separately. We placed "cold" code in the cold cache which is where the snippets would end up. As such we could run into a scenario where the code caches themselves got allocated with a gap of 4GB. They don't necessarily have to be 4GB large. Do I understand this correctly?

**Comment by joransiu** [comment]
I suspect the two aspects of Tiered Code Cache (historically) is breeding some confusion here.

1. The separation of "warm" and "cold' code / blocks from a given method would be still contained within a single code cache (in other words, a given method would never span across multiple code caches).  The "warm" and "cold" parts grow from opposing ends of a code cache.. and eventually converge somewhere in between.   Given code caches are on the order ~MBs, we would never exceed 4GB range for relative long instructions.  That being said, some compare-and-branch instructions have a much smaller range (iirc 16-bits, so +/- 64k), which forced us to use trampolines at the end of the warm code section to address targets in "cold" code (again within the same method).

2.  The other aspect also sometimes referenced as Tiered Code Cache was related to placement of methods into specific code caches.   We used to place hot or higher compilations into a specific code cache, and cold/warm methods into another.   Trampolines for method invocations across code caches became moot since we started reserving a block of contiguous space for code caches.

**Comment by joransiu** [comment]
I agree given the prevalent availability of the relative-load and store instructions, the purpose of a literal pool base pointer has lost much of its relevance.   At the same time, I believe part of the dynamic literal pool optimization itself was to common (and add an extra child) for the literal pool base value.... is that necessary too?  I would propose we drop both the optimization and locking the register if possible.  In the rare scenario where we do need to generate a reference to literal pool entry, we can materialize it.  (i.e. when running on pre-z10 (?) hardware).   

I also think we should leverage IIHF/IILF to materialize constants (and need to support patching of those sequences as necessary).  In general, those instructions beat out a load from lit pool performance wise, and are not susceptible to cache misses.

**Comment by fjeremic** [comment]
> At the same time, I believe part of the dynamic literal pool optimization itself was to common (and add an extra child) for the literal pool base value.... is that necessary too?

For most purposes dynamic literal pool does not do much as far as integral constants go since the transformations are guarded with codegen queries on whether we would like to materialize a constant inline. For floating point operations, z/Architecture in particular lacks relative load instructions and there do not exist floating point immediate instructions. As such all operations on floating point constants require accesses to the literal pool and hence where the dynamic literal pool optimization shines.

I do agree though as a general trend we should move towards using IIHF / IILF to load 32-bit and 64-bit constants even inline.

**Comment by joransiu** [comment]
Good point about float point.  I guess in those cases, Load relative long doesn't work either, so we do need the optimization.   An alternative would be:  IIHF / IILF / LGDR vs a load from literal pool.

**Comment by fjeremic** [comment]
> Good point about float point. I guess in those cases, Load relative long doesn't work either, so we do need the optimization. An alternative would be: IIHF / IILF / LGDR vs a load from literal pool.

Yes we could go the IIHF route there as well. Though explicitly loading the literal pool via loadaddr at the tree level benefits from commining so for using two floating point constants within an extended basic block the current implementation emits 1 load for the literal pool and 2 loads for the floats where as IIHF / IILF route yields 6 instructions. It definitely does not scale as well. Not sure what the performance trade off would be. Only way to know would be to implement both and benchmark.

**Comment by joransiu** [comment]
Not exactly related, but if we are materializing certain common floating point constants for comparison purposes (i.e. NaN, +0, etc checks), the most optimal instruction would actually be TEST DATA CLASS instruction.   We already generate them in certain evaluators (dbits2l, drem), but not generally.  That avoids even having an entry in litpool for the float const.
